### PR TITLE
Adds text to holders

### PR DESCRIPTION
**Pull request by RaphaelAlvez**
as requested in #407

Text goes into the center of the holder but that can be changed in CSS as usual

image

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3921/pr-test (or any other room on that server)

**Comment by 96LawDawg** [comment]
This needs the accompanying JSON Editor fix.  Right now, there is no button to add text in a holder in the editor and the text label still shows up as orange.

Also, I'm not sure about the defaults of:
  ```
 display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
```
That treats the text in the holder like it is a button.  I think maybe it should be more like a label or a card.  The reason is because I know nothing about css and it took me forever to figure out that to align the text to the left, I had to change display from flex to block.  That step is not necessary in labels or cards and I think text manipulation will be more accessible to the average user if we make the defaults more like those.  Although I would note that both of those are slightly different from each other and different still from a basic widget.  I had no idea we were all over the place on those defaults.

**Comment by ArnoldSmith86** [comment]
For the JSON editor, it just needs to be added to the defaults at the top of holder.js.

**Comment by RaphaelAlvez** [comment]
> This needs the accompanying JSON Editor fix. Right now, there is no button to add text in a holder in the editor and the text label still shows up as orange.
> 
> Also, I'm not sure about the defaults of:
> 
> ```
> display: flex;
> justify-content: center;
> align-items: center;
> text-align: center;
> ```
> 
> That treats the text in the holder like it is a button. I think maybe it should be more like a label or a card. The reason is because I know nothing about css and it took me forever to figure out that to align the text to the left, I had to change display from flex to block. That step is not necessary in labels or cards and I think text manipulation will be more accessible to the average user if we make the defaults more like those. Although I would note that both of those are slightly different from each other and different still from a basic widget. I had no idea we were all over the place on those defaults.

Which default do you think looks better?

image

**Comment by RaphaelAlvez** [comment]
> For the JSON editor, it just needs to be added to the defaults at the top of holder.js.

done

**Comment by ArnoldSmith86** [comment]
The top. But adding a little margin to the middle one would probably also work.

**Comment by 96LawDawg** [comment]
I prefer the middle one.  But I also think it should be similar to cards for consistency.  Holder hold cards.  Text formatting between them should be easy for the casual user.

**Comment by rogerl50** [comment]
I prefer the top one.

**Comment by bjalder26** [comment]
I think the top one looks best, but my question would be which one has the most utility.  I have not had a desire to do this so I am not the best person to decide that, but I would think the most common use would be for the text to peak above the cards in a holder.  However, maybe I'm wrong. 

image

**Comment by RaphaelAlvez** [comment]
> The top. But adding a little margin to the middle one would probably also work.

Added this to the example list

 
image


> But I also think it should be similar to cards for consistency. Holder hold cards.

I disagree with that logic.

I just looked at trading card games playmats and I would say that around 1/2 had text in the middle of the holder, 1/4 had text at the bottom and 1/4 had text on top. bottom and top always had a bit of a margin (padding)
note that with display text you can use align-items for all of these options

image

**Comment by bjalder26** [comment]
> I just looked at trading card games playmats and I would say that around 1/2 had text in the middle of the holder, 1/4 had text at the bottom and 1/4 had text on top. bottom and top always had a bit of a margin (padding) note that with display text you can use align-items for all of these options

I think this is a really good point, because it kind of shows expectations and preference.  However, one think that I would like to point out is that unlike a playmat, we have the ability to strictly enforce the location of cards in a holder, so a top or bottom text might become more advantageous.  I'm just guessing at what people might want though, so if seeing the text above or below the card is not desirable, then I agree that vertically centered looks best to me.

**Comment by RaphaelAlvez** [comment]
As discussed privately with @96LawDawg I added textPosition with options for top, bottom and center (center is a lie but I think it might be useful to already have a default value there)

examples and equivalent CSS are shown here http://212.47.248.129:3921/textPosition

**Comment by 96LawDawg** [comment]
Looks good.  Should add the text-position property to the JSON editor.

**Comment by RaphaelAlvez** [comment]
> Looks good.  Should add the text-position property to the JSON editor.

It should be way at the bottom

**Review by ArnoldSmith86 — COMMENTED** [review]
How sure are you that `display: flex` doesn't cause all sorts of problems? Remember that everything in the holder is an actual child of the holder div.

**Comment by RaphaelAlvez** [comment]
> How sure are you that `display: flex` doesn't cause all sorts of problems? Remember that everything in the holder is an actual child of the holder div.

I did some tests. Look at the cards

**Comment by robartsd** [comment]
> How sure are you that `display: flex` doesn't cause all sorts of problems? Remember that everything in the holder is an actual child of the holder div.

Widgets are all `position: absolute` so they are completely out of the flow; they are not affected by the css flex box properties of their parent in any way.

File: client/js/widgets/holder.js [hunk 1]
```diff
@@ -40,6 +50,25 @@ class Holder extends Widget {
     });
   }
 
+  css() {
+    let css = super.css();
+
+    if(this.get('textPosition'))
+      if (this.get('textPosition')=="top"){
+        css += '; align-items: start; padding-top: 10px';
```
Comment (robartsd) [inline client/js/widgets/holder.js:58]: This padding will cause the position of child widgets to change.

**Comment by 96LawDawg** [comment]
This is a bit more complicated than ideal, but it works and it gives more flexibility than this PR would.  I would close this PR.  The technique @robartsd explained above is being added to a tutorial.

**Pull request closed without merging**